pkg/network/dns_snooper_test.go

```diff
@@ -268,28 +268,58 @@ func getKey(
 	}
 }
 
+func hasKeys(keys ...DNSKey) func(map[DNSKey]map[string]DNSStats) bool {
+	return func(stats map[DNSKey]map[string]DNSStats) bool {
+		for _, key := range keys {
+			if stats[key] == nil {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func atLeast(min int) func(map[DNSKey]map[string]DNSStats) bool {
+	return func(stats map[DNSKey]map[string]DNSStats) bool {
+		return len(stats) >= min
+	}
+}
+
 func getStats(
 	snooper *SocketFilterSnooper,
-	expectedCount int,
+	waitUntilFn func(map[DNSKey]map[string]DNSStats) bool, // wait until this function evaluates to true
 ) map[DNSKey]map[string]DNSStats {
 	// DNS timeout is set to 1 second for the tests.
 	// So a 3-second timeout here should provide enough time for an unanswered DNS query to be considered as a timeout.
 	timeout := time.After(3 * time.Second)
+	all := make(map[DNSKey]map[string]DNSStats)
 Loop:
 	// Wait until DNS stats becomes available
 	for {
 		select {
 		case <-timeout:
 			break Loop
 		default:
-			// Break if we have processed all the expected responses
-			if snooper.successes+snooper.errors >= int64(expectedCount) {
-				break Loop
+			// accumulate result
+			last := snooper.GetDNSStats()
+			for key, statsByDomain := range last {
+				for domain, stats := range statsByDomain {
+					if all[key] == nil {
+						all[key] = make(map[string]DNSStats)
+					}
+					all[key][domain] = stats
```

Comment:
There is a chance the same DNSKey/domain combination is returned from two subsequent calls `GetDNSStats()`.   I think we could combine the stats in that case, but I worry about putting  too much logic into this helper.

Reply:
good point. I'm actually hitting this, and wondering if there is a better approach. thoughts?
I think there is not a good solution for this TBH
one option is to add a method to `statsKeeper` that returns a deep copy of the data

Reply:
Maybe we're talking about different issues? I'm not sure what a deep copy would fix.   

I was thinking that we would wrap all the 'racy' test criteria in assert.Eventually.  It makes it clearer what each test is doing at the expense of extra nesting.

Reply:
oh, so what I mean is: every time you call `GetStats()` you reset all the tests. This is why I don't think `assert.Eventually` is a good idea here, because you would likely get partial results each time you evaluate. That's why I'm manually merging things.

If we do a deep copy of the stats data (without resetting it), we can call `getDNSStatsCopy` multiple times until the conditions are true and without having to merge stuff manually.
In that case we could also use `require.Eventually`, but in any case I think something like `getDNSStatsCopy` is a requirement too

Reply:
Gotcha, so `getDNSStatsCopy` would return the cumulative stats. 

I'm not a huge fan of having a method like that since it's only going to be used under testing and I (think) it would require a bunch of new logic in stat_keeper.

One thing we could do with eventually is have the eventually callback do its own test-specific aggregation. Something like:


```
var matchedRequests int
assert.Eventually(func(t *testing.T) {
    matchedRequests += len(getStats(...)) 
    return matchedReqests == 2
})
```